rails/README.md

```diff
@@ -34,10 +34,93 @@ A guide for building great Rails apps.
+database with a given group of work, write a [strong migration].
 - Don't change a migration after it has been merged into `main` if the desired
   change can be solved with another migration.
+- Schema migrations and data migrations should be kept separate.
+- We use the [`strong_migrations`], [`fx`], and [`data_migrate`] gems, along
+  with a 3 phase approach for any change where we are renaming or moving columns
+  around. These phases should happen over 3 different releases.
+  - Phase 1:
+    - _Code_ - Add deprecation warnings for old attribute.
+    - _Function_ Create function to duplicate data from old to new column.
+    - _Trigger_ - Create trigger to run function on old column.
+    - _Schema Migration_ - Create new column in the appropriate table.
+    - _Schema Migration_ - Add phase 1 trigger and function.
+    - _Data Migration_ - Create a backfill from old to new column.
+  - Phase 2:
+    - _Code_ - Move reads from old to new column.
+    - _Function_ Create function to duplicate data from new to old column.
+    - _Trigger_ - Create trigger to run function on new column.
+    - _Schema Migration_ - Remove phase 1 function and trigger.
+    - _Schema Migration_ - Add phase 2 trigger and function.
+  - Phase 3:
+    - _Code_ - Remove deprecation warnings for old attribute.
+    - _Functions and Triggers_ - Remove all function and trigger.
+    - _Schema Migration_ - Remove phase 2 function and trigger.
+    - _Schema Migration_ - Remove old column.
+
+### Functions
+
+- These are custom function that can later be called via SQL.
+- Stored in `db/functions`.
+- Added to the database via `create_function` schema migrations.
+- Removed from the database via `drop_function` schema migrations.
+- Written in `plpgsql`, [SQL Procedural Language] for PostgresSQL.
+
+### Triggers
+
+- These are event [triggers] that fire when the given event happens.
+- Stored in `db/triggers`.
+- Added to the database via `create_trigger` schema migrations.
+- Removed from the database via `drop_trigger` schema migrations.
+
+### Schema Migrations
+
+- These are standard rails migrations, they are stored in `db/migrate`.
+- No data should be migrated in these.
+- Since there are no data, there should be no references to models.
 - If there are default values, set them in migrations.
-- Use SQL, not `ActiveRecord` models, in migrations.
 - [Add foreign key constraints] in migrations.
+- Add appropriate indexes.
+
+### Data Migrations
+
+- Use SQL, not `ActiveRecord` models, in data migrations.
+- Depending on need data migrations fall into three use cases as follows.
+
+#### Data Migrate Gem
+
+- We use the [`data_migrate`] gem for managing data migrations.
+- This should be the default data migration method.
+- These are run interwoven with the schema migrations.
+- They are stored in `db/data`.
+- Use the generator `rails g data_migration some_new_data_migration`.
+- Make them reversible in case we have to rollback.
```

Comment:
I am tempted to say that these should error our if we can't rollback. What do others think?